adafruit_requests.py

```diff
@@ -37,35 +37,117 @@
+            the data received. The maximum amount of data to be received at once is specified
+            by bufsize. The meaning of the optional flags kwarg is implementation-specific."""
+            ...
+
+    class SupportsRecvInto(Protocol):
+        """Describes a type that possesses a socket recv_into() method."""
+
+        def recv_into(
+            self, buffer: bytearray, nbytes: int = ..., flags: int = ...
+        ) -> int:
+            """Receive up to nbytes bytes from the socket, storing the data into the provided
+            buffer. If nbytes is not specified (or 0), receive up to the size available in the
+            given buffer. The meaning of the optional flags kwarg is implementation-specific.
+            Returns the number of bytes received."""
+            ...
+
+    class CircuitPythonSocketType(
+        CommonCircuitPythonSocketType,
+        SupportsRecvInto,
+        SupportsRecvWithFlags,
+        Protocol,
+    ):  # pylint: disable=too-many-ancestors
+        """Describes the structure every modern CircuitPython socket type must have."""
+
+        ...
+
+    class StandardPythonSocketType(
+        CommonSocketType, SupportsRecvInto, SupportsRecvWithFlags, Protocol
+    ):
+        """Describes the structure every standard Python socket type must have."""
 
-try:
-    from typing import Union, TypeVar, Optional, Dict, Any, List, Type
-    import types
-    from types import TracebackType
-    import ssl
-    import adafruit_esp32spi.adafruit_esp32spi_socket as esp32_socket
-    import adafruit_wiznet5k.adafruit_wiznet5k_socket as wiznet_socket
-    import adafruit_fona.adafruit_fona_socket as cellular_socket
-    from adafruit_esp32spi.adafruit_esp32spi import ESP_SPIcontrol
-    from adafruit_wiznet5k.adafruit_wiznet5k import WIZNET5K
-    from adafruit_fona.adafruit_fona import FONA
-    import socket as cpython_socket
-
-    SocketType = TypeVar(
-        "SocketType",
-        esp32_socket.socket,
-        wiznet_socket.socket,
-        cellular_socket.socket,
-        cpython_socket.socket,
-    )
-    SocketpoolModuleType = types.ModuleType
-    SSLContextType = (
-        ssl.SSLContext
-    )  # Can use either CircuitPython or CPython ssl module
-    InterfaceType = TypeVar("InterfaceType", ESP_SPIcontrol, WIZNET5K, FONA)
+        def connect(self, address: Union[Tuple[Any, ...], str, bytes]) -> None:
+            """Connect to a remote socket at the provided address."""
+            ...
+
+    SocketType = Union[
+        LegacyCircuitPythonSocketType,
+        CircuitPythonSocketType,
+        StandardPythonSocketType,
+    ]
+
+    SocketpoolModuleType = ModuleType
+
+    class InterfaceType(Protocol):
+        """Describes the structure every interface type must have."""
+
+        @property
+        def TLS_MODE(self) -> int:  # pylint: disable=invalid-name
+            """Constant representing that a socket's connection mode is TLS."""
+            ...
```

Comment:
I discovered CircuitPython does not appear to support class variables, because previously I wrote this `Protocol` class using this more concise form:

```python
class InterfaceType(Protocol):
    """Describes the structure every interface type must have."""
    
    TLS_MODE: int
```
but that resulted in a `SyntaxError: invalid syntax` exception being raised for the `TLS_MODE: int` line.

Luckily `Protocol` supports [an alternative syntax for these kinds of variables (specifically for Python 2.7-3.5)](https://www.python.org/dev/peps/pep-0544/#using-protocols-in-python-2-7-3-5) using `property` which I was able to leverage here.